Issue #12 Complete - style, order, sort list items by frequency

laurenyz opened the pull request:
## Description 
- Items are sorted into four categories by frequency: soon, not-soon, kind-of-soon, inactive.
- Within each group, items are sorted alphabetically
- Groups are differentiated by checkmark background color (green, orange, red, gray)
- aria labels reflect the frequency group

## Related Issue
closes #12
<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->

## Acceptance Criteria
- Items in the list are shown as visually distinct (e.g., with a different background color on the list item) according to how soon the item is expected to be bought again: Soon, Kind of soon, Not soon, Inactive
- Items should be sorted by the estimated number of days until next purchase
- Items with the same number of estimated days until next purchase should be sorted alphabetically
- Items in the different states should be described distinctly when read by a screen reader

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
| ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
<img width="1280" alt="Screen Shot 2020-09-10 at 7 19 44 PM" src="https://user-images.githubusercontent.com/50968795/92831357-4f5e7d00-f3a4-11ea-876f-50382b397400.png">

### After
<img width="1280" alt="Screen Shot 2020-09-10 at 7 24 29 PM" src="https://user-images.githubusercontent.com/50968795/92831426-60a78980-f3a4-11ea-92ad-3c377422fc43.png">

## Testing Steps / QA Criteria
1. Use token `film stag lipid`
2. Notice that all four seconds (soon, kind of soon, not soon, inactive) are differentiated by color, and are sorted alphabetically within each section.
3. You can then check the frequency for each item in firebase.

QMaximillian committed:
Sort shopping list items

Sort shopping list items by estimated number of days and alphabetically within each range

laurenyz committed:
added classes for each item depending on purchase date

QMaximillian committed:
Remove extra sort, add aria-label, adjust sort order

Remove redundant frequency sort on sortedResults function, add aria-label to List.js li elements in order to make sure screen-readers read out correct purchase date frequency color-coded enums, adjust returned array in sortedResults function to return an alphabetically sorted array based on frequency ranges

laurenyz committed:
added background colors to checkboxes

laurenyz committed:
added css for checkmarks so that they display various colors depending on frequency, and added comments explaining our logic in List

laurenmbeatty commented:
👋  @QMaximillian and @laurenyz I'm not seeing the checkboxes:

![image](https://user-images.githubusercontent.com/10378331/92952409-89239800-f425-11ea-958a-72a1f6d0b682.png)

QMaximillian commented:
> 👋 @QMaximillian and @laurenyz I'm not seeing the checkboxes:
> 
> ![image](https://user-images.githubusercontent.com/10378331/92952409-89239800-f425-11ea-958a-72a1f6d0b682.png)

Hmm, I just checked the branch and pulled down the most recent changes and the checkboxes are there. I'm wondering if maybe this is a bug? Maybe @laurenyz and I can try and reproduce.

laurenmbeatty commented:
Oh! it's the opacity of 1%...if I get rid of it then I can see them:

![image](https://user-images.githubusercontent.com/10378331/92953565-8b86f180-f427-11ea-91e1-a87bcfad619d.png)

I think those values need to be 0.0 - 1.0. So for `40%` do `0.4`

katmdev commented on src/App.css:
```diff
@@ -56,6 +56,113 @@ ul {
-.checkbox {
+/* .checkbox {
   margin: 10px;
+} */
```
Should this and the commented out content at lines 136 - 139 be removed?

katmdev commented on src/components/List.js:
```diff
@@ -19,6 +19,51 @@ You cannot undo this action, and this item's purchase history will be lost.`,
+    let notSoon = [];
+    let inactive = [];
+
+    results.forEach(result => {
+      //separates each item into arrays by frequency
+      switch (true) {
+        case result.frequency <= 604800000: //milliseconds in 7 days
+          result.timeClass = 'soon';
+          soon.push(result);
+          break;
+        case result.frequency <= 1209600000: //milliseconds in 14 days
+          result.timeClass = 'kind-of-soon';
+          kindOfSoon.push(result);
+          break;
+        case result.frequency <= 2592000000: //milliseconds in 30 days
+          result.timeClass = 'not-soon';
+          notSoon.push(result);
+          break;
+        default:
+          result.timeClass = 'inactive'; //everything above else defaults to inactive
+          inactive.push(result);
+          break;
+      }
```
Love the use of a switch statement! As the intention of this feature is to show how soon an item needs to be purchased, calculating frequency + last purchase date - current date may be more accurate, as the frequency value only indicates how often the item is purchased.

laurenyz replied:
Hi Kat!  We were actually very unsure of how to handle this.  We know that it requested sorting by upcoming purchase date, but what happens when the item is past due?  How long should it be allowed to sit there?  The other really confusing thing is when these items are added, we put them in the categories of soon, kind-of-soon, and not-soon, which reflects their frequency.  Then, we're asked to sort them into groups like that again, but it's not necessarily reflecting the frequency - so an item added as "not soon" might show up as "soon" if it's coming up in 3 days, and that might be confusing to someone looking at labels and thinking they put that as not soon.  Finally, it says that items should be sorted alphabetically only if their days are the exact same, but it probably looks better to do alphabetically in one category.  If we do it just if they have the same date, I don't think it will look organized.  It might look like "  carrots, kiwi, apple, cheese" if carrots and kiwis are coming up sooner than the others.  

This was a little stream-of-consciousness, but please let us know if you have specific ideas because it was very confusing how we should handle that part!

katmdev replied:
> We know that it requested sorting by upcoming purchase date, but what happens when the item is past due? How long should it be allowed to sit there?

I think this is handled by having it show as `soon` until 2 * frequency past its last purchase date, and then it moves to inactive: 

> - Inactive (when there’s only 1 purchase in the database or the purchase is really out of date [the time that has elapsed since the last purchase is 2x what was estimated])

> When these items are added, we put them in the categories of soon, kind-of-soon, and not-soon, which reflects their frequency. Then, we're asked to sort them into groups like that again, but it's not necessarily reflecting the frequency

I think this initial frequency is relative to the date added, not necessarily indicative of a user's perceived purchase frequency, and could be used in tandem with a `dateAdded` value (like what we created in our AC in order to show next purchase date in the details). I think this also speaks to the move to `inactive` once we have one purchase in the purchase history as per the AC (because we don't have enough data to calculate a frequency).

> so an item added as "not soon" might show up as "soon" if it's coming up in 3 days, and that might be confusing to someone looking at labels and thinking they put that as not soon. 

My interpretation was that items added as `not soon` will move through a life cycle of `not soon` -> `kind of soon` -> `soon` -> and then back to `not soon` once the item is purchased or `inactive` if the item is not purchased within the 2 * frequency as per the AC.

> Finally,` it says that items should be sorted alphabetically only if their days are the exact same, but it probably looks better to do alphabetically in one category.

I think that having the items sorted by next purchase date makes more sense from a user perspective when determining how soon an item should be purchased.

QMaximillian committed:
Remove and update css in App.css

Change opacity to use decimal values, eliminate commented out code on line 136-139

laurenmbeatty commented:
I will let you all handle the logic for how you want things sorted. My concern is with the keyboard accessibility of the checkboxes. If i try to use my keyboard to navigate, it doesn't focus on the checkbox, and I can't tell where I should press enter/space to select the checkbox. 

See if you can super-impose the actual input on top of the span that's styled to look like the checkbox. You'd need to give it a position of absolute, and top: 0 and left: 0 (something like that), give it a height and width, and a z-index of 2, or something.

This is not a requirement, but a very nice to have.

This has focus: 
![image](https://user-images.githubusercontent.com/10378331/93007478-f5d68980-f52e-11ea-9303-2acf94466ee8.png)

laurenmbeatty commented on src/components/List.js:
```diff
@@ -19,6 +19,51 @@ You cannot undo this action, and this item's purchase history will be lost.`,
+        default:
+          result.timeClass = 'inactive'; //everything above else defaults to inactive
+          inactive.push(result);
+          break;
```
Nice commenting here ^.

laurenyz committed:
added css prop for input focus

The custom css for the checkmark is essentially hiding the original checkmark and placing another custom checkmark on top. With this added focus prop, the custom checkmark is highlighted when the element is selected (ex with screen reader or by tabbing)

laurenyz committed:
deleted unnecessary comments, but left css for a change in background when checked commented out in case we want to use in the future

laurenyz committed:
merged

laurenyz committed:
adjusted sort functionality to better reflect AC.

Instead of sorting according to frequency, the items are now sorted by days left until predicted next purchase.

laurenyz commented:
@katmdev @laurenmbeatty @QMaximillian @TianaJohnson 
Phew - adjusted css for focus on checkmarks.  Also update sort function so that it is now sorted/color coded according to days left until predicated purchase date.  I currently left two numbers in the list item so that it's easier for you guys to check.  The first number is the number of days left until anticipated purchase, and the second number is the frequency.  We still need to hook up our branch to the other branch, but there's currently a placeholder date in the event that an item doesn't have any purchase dates.  Let me know if you have any questions!

laurenmbeatty commented on src/App.css:
```diff
@@ -166,3 +166,7 @@ ul {
+
+.container input:focus ~ .checkmark {
+  box-shadow: 0 0 0 3px #5e9ed6;
+}
```
Excellent!

laurenmbeatty commented on src/components/List.js:
```diff
@@ -66,23 +138,34 @@ You cannot undo this action, and this item's purchase history will be lost.`,
                   key={result.id}
-                  className={checkTime(time) ? 'deactivated' : null}
+                  className={checkTime(time) ? `deactivated` : null}
+                  // aria-label={result.timeClass.split('-').join(' ')}
```
I would put the aria-label on the input.

mjhcodes commented on src/App.css:
```diff
@@ -56,6 +56,108 @@ ul {
+  background-color: gray;
+  opacity: 0.4;
+  cursor: pointer;
+}
```
Not a big deal, but there's a good amount of repetitive code here between the various frequency class selectors. I'd be curious if it'd be possible to create one class selector for the _.checkmark_, which holds the bulk of the properties (e.g. `position`, `top`, `left`, `height`, etc.) and then also create the individual class selectors for each frequency itself, which would hold their own unique `background-color` property only.

This way, if anything needs to be changed to one of the _.checkmark_ properties, you'd only need to change it once, as opposed to four separate places.

mjhcodes commented on src/components/List.js:
```diff
@@ -19,6 +19,78 @@ You cannot undo this action, and this item's purchase history will be lost.`,
+        ),
+      //just sorts inactive items alphabetically. It doesn't make as much sense to sort HOW inactive an item is
+      ...inactive.sort((a, b) =>
+        a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1,
```
Very simplified sorting approach. I like it! :-)

mjhcodes commented on src/components/List.js:
```diff
@@ -19,6 +19,78 @@ You cannot undo this action, and this item's purchase history will be lost.`,
+  function sortedResults() {
+    let active = [];
+    let inactive = [];
+    results.forEach(result => {
+      //pulls out inactive items - identified by having a negative number less than the negative frequency, or has a frequency greater than 30.
+      //for example, if an item has a frequency of 7 days, and the anticpated days left is -8, that means it's been 15 days since the last purchase date, and is inactive
+      if (
+        predictedNextPurchase(lastPurchaseDate(result), result.frequency) <
+          -1 * Math.round(result.frequency / 86400000) ||
+        Math.round(result.frequency / 86400000) > 30
+      ) {
+        result.timeClass = 'inactive';
+        inactive.push(result);
```
The inactive AC wasn't part of the issue when I participated in the program, so really cool to see your approach to tackling this one. Great job!

mjhcodes reviewed (approved):
Checks all the colored boxes! 😋  Great all-around approach to sorting and I like the choice for how to distinguish between item frequency. Happy to approve at this stage, as I understand the numbers currently displayed will be removed prior to merging. Great work! 👏

QMaximillian committed:
Refactoring, moving aria-label to input, add addedDate property to results

Add addedDate property all results from querySnapshot in App.js, set predictedNextPurchase function value to a local variable to prevent repeated function calls in switch statement, move aria-label from li element to input element, replace placeholder value for addedDate in lastPurchaseDate function

QMaximillian committed:
Create individual checkmark class to modularize css for .checkmark class in List.css
